core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java

```diff
@@ -13,22 +13,59 @@
  */
 package io.trino.spi.classloader;
 
-import java.io.Closeable;
+import java.util.function.Supplier;
 
 public class ThreadContextClassLoader
-        implements Closeable
 {
-    private final ClassLoader originalThreadContextClassLoader;
+    private ThreadContextClassLoader() {}
 
-    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
+    public static <T> T withClassLoader(ClassLoader newThreadContextClassLoader, Supplier<T> supplier)
     {
-        this.originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
         Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
     }
 
-    @Override
-    public void close()
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
     {
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
+    }
+
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
+            extends Supplier<T>
+    {
+        default T get()
+        {
+            try {
+                return supplyOrThrow();
+            }
+            catch (Exception e) {
+                // TODO(sergey-melnychuk) There must be a better way to address this
```

Comment:
Not sure how exactly, but I have a strong feeling that this can be done better than just wrapping into `RuntimeException`. Suggestions?